foundry/auto_self_nominate.rs

```diff
@@ -0,0 +1,193 @@
+                let account = self.account_provider.get_unlocked_account(&address)?;
+                account.sign(&hash)?
+            }
+        };
+        Ok(result)
+    }
+
+    pub fn address(&self) -> Option<&Address> {
+        self.signer.as_ref().map(|(address, _)| address)
+    }
+}
+
+pub struct AutoSelfNomination {
+    client: Arc<dyn ConsensusClient>,
+    signer: SelfSigner,
+}
+
+impl AutoSelfNomination {
+    pub fn new(client: Arc<dyn ConsensusClient>, ap: Arc<AccountProvider>, address: Address) -> Arc<Self> {
+        Arc::new(Self {
+            client,
+            signer: SelfSigner::new(ap, address),
+        })
+    }
+
+    pub fn send_self_nominate_transaction(&self, matches: &ArgMatches) {
+        let config = load_config(matches).unwrap();
+        let account_address = config.mining.engine_signer.unwrap();
+        let defualt_metadata = config.mining.self_nomination_metadata.unwrap();
+        let target_deposite = config.mining.self_target_deposit.unwrap();
+        let interval = config.mining.self_nomination_interval.unwrap();
+        let self_client = self.client.clone();
+        let self_signer = self.signer.clone();
+        thread::Builder::new()
+            .name("Auto Self Nomination".to_string())
+            .spawn(move || loop {
+                AutoSelfNomination::send(
+                    &self_client,
+                    &self_signer,
+                    &account_address,
+                    &defualt_metadata,
+                    target_deposite,
+                );
+                thread::sleep(Duration::from_millis(interval));
+            })
+            .unwrap();
+    }
+
+    fn send(
+        client: &Arc<dyn ConsensusClient>,
+        signer: &SelfSigner,
+        account_address: &PlatformAddress,
+        metadata: &str,
+        targetdep: u64,
+    ) {
+        let metabytes = metadata.rlp_bytes();
+        let mut dep = targetdep;
+        let address = account_address.address();
+        let block_id = BlockId::Latest;
+        let state = client.state_at(block_id).unwrap();
+        let banned = Banned::load_from_state(&state).unwrap();
+        if banned.is_banned(address) {
+            cwarn!(ENGINE, "Account is banned");
+            return
+        }
+        let jailed = Jail::load_from_state(&state).unwrap();
+        let prisoner = jailed.get_prisoner(&address).unwrap();
+        let current_term = client.current_term_id(block_id).unwrap();
+        if prisoner.custody_until <= (current_term) {
+            cwarn!(ENGINE, "Account is still in custody");
+            return
+        }
+        let remaining = prisoner.released_at.min(current_term);
+        let candidate = Candidates::load_from_state(&state).unwrap();
+        let candidate_need_nomination = candidate.get_candidate(&address).unwrap();
+
+        if candidate_need_nomination.nomination_ends_at <= current_term + NEED_NOMINATION_UNDER_TERM_LEFT {
+            cwarn!(ENGINE, "No need self nominate");
```

Comment:
Let's make this log message to use "cdebug!". CodeChain will log this message every few minutes. There will be too many logs. Also, no meaning to show this message to the user.